File: debuglogger/src/main/java/io/github/japskiddin/debuglogger/DebugLogger.java
package io.github.japskiddin.debuglogger;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import io.github.japskiddin.debuglogger.databinding.DebugLoggerBinding;
import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("Convert2Lambda") public class DebugLogger extends LinearLayout implements
    DefaultLifecycleObserver {
  private final static long HANDLER_DELAY = 2000L;
  private final Handler logHandler = new Handler(Looper.getMainLooper());
  private LogAdapter logAdapter;
  private DebugLoggerBinding binding;

  public DebugLogger(Context context) {
    super(context);
    ((LifecycleOwner) context).getLifecycle().addObserver(this);
  }

  public DebugLogger(Context context, @Nullable AttributeSet attrs) {
    super(context, attrs);
    ((LifecycleOwner) context).getLifecycle().addObserver(this);
  }

  private final Runnable debugLogRunnable = new Runnable() {
    @Override public void run() {
      List<LogEvent> logs = new ArrayList<>(LogManager.getInstance().getLogs());
      if (LogManager.getInstance().isEnabled() && logs.size() > 0) {
        for (LogEvent log : logs) {
          logAdapter.addItem(log);
        }
        binding.rvLogs.scrollToPosition(logAdapter.getItemCount() - 1);
        LogManager.getInstance().clear();
      }
      logHandler.postDelayed(this, HANDLER_DELAY);
    }
  };

  private void onLogsClear() {
    LogManager.getInstance().clear();
    logAdapter.clear();
  }

  private void onCopyLogs() {
    ClipboardManager clipboard =
        (ClipboardManager) getContext().getSystemService(Context.CLIPBOARD_SERVICE);
    ClipData clip = ClipData.newPlainText("Copied text", logAdapter.getAllText());
    clipboard.setPrimaryClip(clip);
    Toast.makeText(getContext(), "Text copied", Toast.LENGTH_LONG).show();
  }

  @Override public void onCreate(@NonNull LifecycleOwner owner) {
    binding = DebugLoggerBinding.inflate(LayoutInflater.from(getContext()), this, true);
    logAdapter = new LogAdapter();
    binding.rvLogs.setAdapter(logAdapter);
    binding.btnLogClear.setOnClickListener(new OnClickListener() {
      @Override public void onClick(View v) {
        onLogsClear();
      }
    });
    binding.btnLogCopy.setOnClickListener(new OnClickListener() {
      @Override public void onClick(View v) {
        onCopyLogs();
      }
    });
  }

  @Override public void onStart(@NonNull LifecycleOwner owner) {
  }

  @Override public void onStop(@NonNull LifecycleOwner owner) {
  }

  @Override public void onPause(@NonNull LifecycleOwner owner) {
    logHandler.removeCallbacks(debugLogRunnable);
  }

  @Override public void onResume(@NonNull LifecycleOwner owner) {
    logHandler.post(debugLogRunnable);
  }

  @Override public void onDestroy(@NonNull LifecycleOwner owner) {
    logHandler.removeCallbacks(debugLogRunnable);
    binding = null;
  }
}

File: debuglogger/src/main/java/io/github/japskiddin/debuglogger/ui/DebugLogger.kt
package io.github.japskiddin.debuglogger.ui

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.LinearLayout
import android.widget.Toast
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import io.github.japskiddin.debuglogger.adapter.LogAdapter
import io.github.japskiddin.debuglogger.databinding.DebugLoggerBinding
import io.github.japskiddin.debuglogger.manager.LogManager

class DebugLogger : LinearLayout, DefaultLifecycleObserver {
    private lateinit var binding: DebugLoggerBinding
    private lateinit var logAdapter: LogAdapter
    private val logHandler = Handler(Looper.getMainLooper())

    constructor(context: Context) : this(context, null, 0)

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)

    constructor(context: Context, attrs: AttributeSet?, defaultStyleAttr: Int) : super(
        context,
        attrs,
        defaultStyleAttr
    ) {
        (context as LifecycleOwner).lifecycle.addObserver(this)
    }

    private val debugLogRunnable = object : Runnable {
        override fun run() {
            if (LogManager.getInstance().isEnabled()) {
                logAdapter.submit(LogManager.getInstance().getLogs())
            }
            logHandler.postDelayed(this, HANDLER_DELAY)
        }
    }

    private fun onLogsClear() {
        LogManager.getInstance().clear()
        logAdapter.clear()
    }

    private fun onCopyLogs() {
        val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val clip = ClipData.newPlainText("Copied text", logAdapter.getAllItemsString())
        clipboard.setPrimaryClip(clip)
        Toast.makeText(context.applicationContext, "Text copied", Toast.LENGTH_LONG).show()
    }

    override fun onCreate(owner: LifecycleOwner) {
        binding = DebugLoggerBinding.inflate(LayoutInflater.from(context), this, true)
        logAdapter = LogAdapter { binding.rvLogs.scrollToPosition(logAdapter.itemCount - 1) }
        binding.rvLogs.adapter = logAdapter
        binding.btnLogClear.setOnClickListener { onLogsClear() }
        binding.btnLogCopy.setOnClickListener { onCopyLogs() }
    }

    override fun onPause(owner: LifecycleOwner) {
        logHandler.removeCallbacks(debugLogRunnable)
    }

    override fun onResume(owner: LifecycleOwner) {
        logHandler.post(debugLogRunnable)
    }

    override fun onDestroy(owner: LifecycleOwner) {
        logHandler.removeCallbacks(debugLogRunnable)
    }

    companion object {
        private const val HANDLER_DELAY = 2000L
    }
}